---
title: "Metagenomic data analysis pipeline"
subtitle: "For Oxford Nanopore Technologiesâ€™ MinION long-read, single-end shotgun sequencing data"
author: "Moreno Serafino"
date: "`r Sys.Date()`"
output:
  html_document: default
  word_document: default
  pdf_document: default
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```
<br><br>

--------------------------------------------------
--------------------------------------------------

# Introduction

This pipeline is designed for analyzing Next-Generation Sequencing data. The raw data used for testing this pipeline originate from two separate ONT MinION long-read, single-end shotgun DNA-sequencing experiments on ditch water using R9.4.1 flow-cells, presented in fastq format. This Markdown document will therefore use the ditch water experiment data to guide the reader through the analysis pipeline. The point of this pipeline, however, is that data from any other similar sequencing experiment can be used as input. The end result will be a comprehensive overview of microbe composition in origin samples, supported by multiple instances of quality control. 

Note: this markdown document is aimed at people with limited to zero command line experience.

Note: phrases in **bold** specify commands; phrases in *italics* specify scripts.
<br>

### Contents

[inhoudsopgave]
<br>

### Environment setup:

To properly run this pipeline, a number of 'packages' needs to be installed first. These packages are available online and contain the code necessary to execute all downstream scripts and commands. The 'conda' package management system is used to install these packages to a specific 'conda environment', which will help avoid conflicts with other programs present on the server.

**conda update** updates the conda system; it is good practice to update first.<br>
**conda create [name_of_environment]** will create a new environment.<br>
**conda activate [name_of_environment]** will switch the active working directory to that environment and thus activate all installed packages.<br>
**conda install [name_of_package]** will install packages to the currently active conda environment.

For this pipeline, the 'meta' conda environment is created and configured as follows:


```{bash, echo=TRUE, eval=FALSE}

# update conda; 'echo "y"' will automatically answer a confirmation prompt:

echo "y" | conda update -n base conda

# create meta environment:

conda create meta

# activate conda meta environment:

conda activate meta

# install all packages necessary for downstream scripts and commands:

echo "y" | conda install fastqc
echo "y" | conda install Porechop
echo "y" | conda install [...]

```

Note: The currently active conda environment is always displayed in the terminal between brackets at the start of each line.
<br>

### Library setup:

[blabla]

```{r}

# load libraries:
# ...

```
<br>

### Working directory

The working directory with all the necessary files can be found here: home/1790915/Metagen_SW. Scripts will not work from other directories unless the path is specified. Change your directory with **cd /home/data/1790915/Metagen_SW**. Then check the current working directory with **pwd**. The virtual workspace is now properly set up.
<br>

### About scripts

The next section contains the first bash script. These scripts can simply be run from the terminal. However, most scripts need to know where to find their input data, and where to store their output file(s), among other variables. The scripts are designed so that the user may provide these variables as options. Use **bash [script_name.sh] -h** to print an overview of a script's function and usage. This works for all bash scripts down the line.<br><br><br>

--------------------------------------------------
--------------------------------------------------

# Raw sequencing data overview

The raw data from the ditch water project can be found on the HU server: /home/data/projecticum/SW/raw_data. There are 288 separate .fastq.gz files. For the purposes of testing the pipeline, a subset of 10 files is created. The following files are chosen through random number generation:

fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_17_0.fastq.gz
fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_29_0.fastq.gz
fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_56_0.fastq.gz
fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_78_0.fastq.gz
fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_88_0.fastq.gz
fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_115_0.fastq.gz
fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_173_0.fastq.gz
fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_202_0.fastq.gz
fastq_runid_e0b226454119dd745e929660cd7d0704fc8084f2_16_0.fastq.gz
fastq_runid_e0b226454119dd745e929660cd7d0704fc8084f2_40_0.fastq.gz

These files are copied from the server, and concatenated (combined) into one .fastq.gz file:

```{bash, echo=TRUE, eval=FALSE}

# copy the specified files from the server to the raw_data folder:

cp /home/data/projecticum/SW/raw_data/fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_17_0.fastq.gz ../raw_data
cp /home/data/projecticum/SW/raw_data/fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_29_0.fastq.gz ../raw_data
cp /home/data/projecticum/SW/raw_data/fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_56_0.fastq.gz ../raw_data
cp /home/data/projecticum/SW/raw_data/fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_78_0.fastq.gz ../raw_data
cp /home/data/projecticum/SW/raw_data/fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_88_0.fastq.gz ../raw_data
cp /home/data/projecticum/SW/raw_data/fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_115_0.fastq.gz ../raw_data
cp /home/data/projecticum/SW/raw_data/fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_117_0.fastq.gz ../raw_data
cp /home/data/projecticum/SW/raw_data/fastq_runid_49c9a95cddfedd94200fbd3b545c9c9571fc43c7_202_0.fastq.gz ../raw_data
cp /home/data/projecticum/SW/raw_data/fastq_runid_e0b226454119dd745e929660cd7d0704fc8084f2_16_0.fastq.gz ../raw_data
cp /home/data/projecticum/SW/raw_data/fastq_runid_e0b226454119dd745e929660cd7d0704fc8084f2_40_0.fastq.gz ../raw_data

# concatenate into one file:

zcat /raw_data/fastq_runid* > combined_10.fastq

# compress to limit disk space:
 
gzip /raw_data/combined_10.fastq

# remove the copied (separate) fastq files:

rm /raw_data/fastq_runid*

```

```{bash, echo=TRUE, eval=TRUE}

# check the combined file's format by viewing part of the first read:

zcat raw_data/combined_10.fastq.gz | head -n4 | cut -c1-400

# it should look something like this:

```
<br>

The first line in the read shows the metadata, specifically the type of sequencing experiment the data originates from (in this case R9.4.1 MinION, DNA). 
If your data does not provide clear metadata, a quick summary report can be created through the *FASTQ_summarizer.sh* script. 

```{bash, echo=TRUE, eval=FALSE}

# print the script's function and usage (optional):

bash FASTQ_summarizer.sh -h

# execute the script with proper flags:

bash FASTQ_summarizer.sh -I raw_data -O analyse/analyse_misc

```

```{bash, echo=TRUE, eval=TRUE}

# open summary report:

cat analyse/analyse_misc/FASTQ_summary.txt

# the report looks like this:

```

The "longest read length" result provides insight in the type of sequencing data: if this result is 300 basecalls or less, it would be a strong indication that the raw data originates from Illumina short-read sequencing instead of long-read sequencing.<br><br><br>

--------------------------------------------------
--------------------------------------------------

# Initial quality control

### FastQC on raw data subset

The FastQC tool will be used to get a quick qualitative overview using the basecall-linked Phred-scores in the dataset. The *FASTQC_reporter.sh* script takes FASTQ data, analyzes the quality scores and creates an html report. The report consists of a number of graphs that will be discussed further. 

```{bash, echo=TRUE, eval=FALSE}

# print the script's function and usage (optional):

bash FASTQC_reporter.sh -h

# execute the script with proper flags:

bash FASTQC_reporter.sh -I raw_data -O analyse/FastQC

```

Open the html report by navigating to the output directory with **cd ~/Metagen_SW/analyse/FastQC**, then click on the .html file and choose 'View in Web Browser'. The "Basic Statistics" section of the FastQC report shows an overview of the metadata. This should correspond with the *FASTQ_summarizer.sh* results from earlier.

![Figure 1: FastQC report on combined_10.fastq.gz](misc/FastQC_combined_10.PNG)

<br>The boxplots in the "Per base sequence quality" graph represent the quality values of the entire input per basecall 'location' in the reads, i.e. the quality scores of the first base of each read combined are represented by the first box. Results of this analysis can be used down the line to trim reads in order to obtain a subset of higher average quality.

Note: FastQC is originally designed for analyzing Illumina or Sanger short read sequencing data. The "Encoding" measurement under "Basic Statistics" should confirm this. Phred scores from short read sequencing are expected to be much higher than those of Nanopore long reads. Phred scores for Nanopore MinION R9.4.1 experiments should preferably be > 8. The color-coding in the "Per base sequence quality" graph should therefore not be taken into account.
<br>

### Porechop adapter trimming

ONT uses certain adapters to assist with sequencing experiments. The 'Porechop' tool is developed to find and excise many of these adapter sequences in fastq files. the Porechop tool can be used as follows:

```{bash, echo=TRUE, eval=FALSE}

# print the porechop description (optional):

porechop -h

# execute porechop with proper flags; note that porechop takes files instead of directories as in- and output:

porechop -i raw_data/combined_10.fastq.gz -o bewerkte_data/combined_10_porechopped.fastq.gz

```

Porechop progress can be viewed in the terminal while the tool is running. After the trimming is complete, check the results by running FastQC on the trimmed data:

```{bash, echo=TRUE, eval=FALSE}

# execute the FastQC_reporter.sh script again, with the trimmed data as input:

bash FASTQC_reporter.sh -I bewerkte_data -O analyse/FastQC

```

Open the new html report and compare the "Per base sequence quality" with the earlier report. 

![Figure 2: FastQC report on combined_10_porechopped.fastq.gz](misc/FastQC_combined_10_porechopped.PNG)

<br>In this example, the low-quality basecalls initially seen in figure 1 have been trimmed off by porechop, increasing the overall quality of the data.





